### PR TITLE
FOGL-8212 Add Housekeeper task to combine a number of threads into one

### PR DESCRIPTION
**Pull request by MarkRiddoch**

### CODE REVIEW

**Review by pintomax — APPROVED** [review]

File: C/common/housekeeper.cpp [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+		m_thread = new thread(hkThread, this);
+	}
+	// Wakeup the thread so that it may schedule the next task
+	m_cv.notify_all();
+}
+
+/**
+ * Remove a registered task fromt eh house keeper queue
+ *
+ * If a task is foudn then the cleanup of the task is called.
```
Comment (RVerhoeff) [inline C/common/housekeeper.cpp:104]: Typo: foudn -> found

File: C/common/housekeeper.cpp [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+		// Start the thread if it is not running
+		m_shutdown = false;
+		m_thread = new thread(hkThread, this);
+	}
+	// Wakeup the thread so that it may schedule the next task
+	m_cv.notify_all();
+}
+
+/**
+ * Remove a registered task fromt eh house keeper queue
```
Comment (RVerhoeff) [inline C/common/housekeeper.cpp:102]: Typo: fromt eh -> from the

File: C/common/housekeeper.cpp [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+ *
+ * If a task is foudn then the cleanup of the task is called.
+ * Note: the task is not deleted by this rotuine and the caller
```
Comment (RVerhoeff) [inline C/common/housekeeper.cpp:105]: Typo: rotuine -> routine

File: C/common/housekeeper.cpp [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+		}
+	}
+	// Wakeup the thread so that it may schedule the next task
+	m_cv.notify_all();
+}
+
+/**
+ * Find the next task to execute. Based on the next execution time
+ * and the priority this routine will get the m_nextTask member
+ * varaible to the next task that should be run.
```
Comment (RVerhoeff) [inline C/common/housekeeper.cpp:131]: Typo: varaible -> variable

File: C/common/housekeeper.cpp [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+	else
+	{
+		m_nextTask = NULL;
+	}
+}
+
+/**
+ * The thread of the housekeeper. This runs until the m_shutdown flag is
+ * set. It looks for tasks to run and exeutes them, waiting on a condition
+ * varaible until there might be a task to run.
```
Comment (RVerhoeff) [inline C/common/housekeeper.cpp:177]: Typo: varaible -> variable

File: C/common/housekeeper.cpp [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+			}
+			lk.lock();
+		}
+
+		// Find the next task to run if any
+		findNextTask();
+
+		if (m_nextTask)
+		{
+			// Wait wait until it is time to run another task
```
Comment (RVerhoeff) [inline C/common/housekeeper.cpp:202]: Typo: wait wait -> wait

File: C/common/include/housekeeper.h [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+#include <logger.h>
+#include <vector>
+#include <thread>
+#include <condition_variable>
+
+class HouseKeeperTask;
+
+/**
+ * A single class that implements a house keeper thread that is used
+ * to run periodic or one-off tasks on a sepreate thread. The purpose
```
Comment (RVerhoeff) [inline C/common/include/housekeeper.h:22]: Typo: sepreate -> separate

File: C/common/include/housekeeper.h [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+ * of the house keeper is to reduce the number of threads that are run
+ * in order to reduce both the meomry footprint and the quanity of threads
```
Comment (RVerhoeff) [inline C/common/include/housekeeper.h:24]: Typo: quanity -> quantity

File: C/common/include/housekeeper.h [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+ * in the system.
+ *
+ * The housekeeper should be used for any task that needs to be run on a
+ * periodic basis but does not require to run for long each time it runs. Likewise
+ * it may be used to run occasional one-off tasks on a seperate thread, again
```
Comment (RVerhoeff) [inline C/common/include/housekeeper.h:29]: Typo: seperate -> separate

File: C/common/include/housekeeper.h [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+		bool			m_shutdown;
+};
+
+/**
+ * Housekeeper priorities used when two house keeper tasks should run
+ * at the same time.
+ */
+typedef enum {
+	/**
+	 * Low priorty task whose execution is not time critical and can
```
Comment (RVerhoeff) [inline C/common/include/housekeeper.h:61]: Typo: priorty -> priority

File: C/common/include/housekeeper.h [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+	 * wait for other tasks to complete.
+	 */
+	HK_LOW,
+	/**
+	 * A normal priority task that should be run if possible on the
+	 * schedule given but that is not critical.
+	 */
+       	HK_NORMAL,
+	/**
+	 * A task that must be run as close as possible to the tiem interval
```
Comment (RVerhoeff) [inline C/common/include/housekeeper.h:71]: Typo: tiem -> time

File: C/common/include/housekeeper.h [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+					m_name(name), m_repeat(repeat), m_priority(priority)
+					{
+						m_nextExecution = time(0) + repeat;
+					};
+		virtual 		~HouseKeeperTask() {};
+		virtual void 		run() = 0;
+		virtual void		cleanup() = 0;
+
+					/**
+					 * Return the next exeuction time for the task
```
Comment (RVerhoeff) [inline C/common/include/housekeeper.h:97]: Typo: exeuction -> execution

File: C/services/common/include/service_handler.h [hunk 13]
```diff
@@ -12,6 +12,9 @@
 #include <config_category.h>
 #include <string>
 #include <management_client.h>
+#include <housekeeper.h>
+
+#define REFRESH_CHECK_INTERVAL	60	// Check the bearer toek every 60 seconds
```
Comment (RVerhoeff) [inline C/services/common/include/service_handler.h:17]: Typo: toek -> token

File: C/services/common/perfmonitor.cpp [hunk 14]
```diff
@@ -155,29 +153,56 @@ void PerformanceMonitor::doCollection(const string& name, long value)
 }
 
 /**
- * The thread that runs to write database values
+ * The hosuekeeper task that runs to write database values
```
Comment (RVerhoeff) [inline C/services/common/perfmonitor.cpp:156]: Typo: hosuekeeper -? housekeeper

File: C/services/common/perfmonitor.cpp [hunk 15]
```diff
@@ -155,29 +153,56 @@ void PerformanceMonitor::doCollection(const string& name, long value)
...
+ *
+ * @param monitor	The performance monitor instance
+ */
+PerformanceTask::PerformanceTask(PerformanceMonitor *monitor) :
+	m_monitor(monitor), HouseKeeperTask("PerformanceMonitor", 60)
+{
+}
+
+/**
+ * The run routine called every 60 seconds by the hosue keeper. Simply
```
Comment (RVerhoeff) [inline C/services/common/perfmonitor.cpp:193]: Typo: hosue -> house

File: C/services/common/perfmonitor.cpp [hunk 16]
```diff
@@ -155,29 +153,56 @@ void PerformanceMonitor::doCollection(const string& name, long value)
...
+ * write the performance counters to the database.
+ */
+void PerformanceTask::run()
+{
+	m_monitor->writeCounters();
+}
+
+/**
+ * The cleanup routine called eithher when the housekeeper shuts down
```
Comment (RVerhoeff) [inline C/services/common/perfmonitor.cpp:202]: Typo: eithher -> either

File: C/services/common/service_security.cpp [hunk 17]
```diff
@@ -92,10 +90,11 @@ bool ServiceAuthHandler::createSecurityCategories(ManagementClient* mgtClient, b
 		m_service_acl = m_mgtClient->getACL(acl_name);
 	}
 
-	// Start thread for automatic bearer token refresh, before expiration
+	// Start hhousekeeper task for automatic bearer token refresh, before expiration
```
Comment (RVerhoeff) [inline C/services/common/service_security.cpp:93]: Typo: hhousekeeper -> housekeeper

File: C/services/common/service_security.cpp [hunk 18]
```diff
@@ -574,155 +573,56 @@ bool ServiceAuthHandler::AuthenticationMiddlewareCommon(shared_ptr<HttpServer::R
 }
 
 /**
- * Refresh the bearer token of the runnign service
- * This routine is run by a thread started in
- * createSecurityCategories.
+ * Refresh the bearer token of the running service
+ * This routine is run by the housekeeper omn a fixed schedule
```
Comment (RVerhoeff) [inline C/services/common/service_security.cpp:577]: Typo: omn -> on

File: C/services/south/ingest.cpp [hunk 19]
```diff
@@ -1499,3 +1490,13 @@ void Ingest::flowControl()
 		m_performance->collect("flow controlled", total);
 	}
 }
+
+/**
+ * Constrcutor for the statistics update housekeeper task
```
Comment (RVerhoeff) [inline C/services/south/ingest.cpp:1495]: Typo: Constrcutor -> Constructor

File: C/common/asset_tracking.cpp [hunk 20]
```diff
@@ -89,22 +80,8 @@ AssetTracker::AssetTracker(ManagementClient *mgtClient, string service)
...
-		this_thread::sleep_for(chrono::milliseconds(10));
-	}
-	if (m_thread)
-	{
-		m_thread->join();
-		delete m_thread;
-		m_thread = NULL;
-	}
+
+	HouseKeeper::getInstance()->addTask(this);
```
Comment (RVerhoeff) [inline C/common/asset_tracking.cpp:84]: I don't understand what this is doing. It looks like _addTask(this)_ is adding this instance of AssetTracker to a task list but this AssetTracker instance is about to vanish. The _this_ pointer will become invalid.

**Review by RVerhoeff — CHANGES_REQUESTED** [review]
Typos plus one question about the AssetTracker destructor.

File: C/common/asset_tracking.cpp (hunk 20 above)
Comment (MarkRiddoch) [reply C/common/asset_tracking.cpp:84]: It should have been removeTask() rather than addTask()

**Review by RVerhoeff — APPROVED** [review]

**Review by pintomax — APPROVED** [review]

**Pull request closed without merging**